pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go

```diff
@@ -345,57 +419,97 @@ func (plugin *kubenetNetworkPlugin) setup(namespace string, name string, id kube
 		plugin.syncEbtablesDedupRules(link.Attrs().HardwareAddr)
 	}
 
-	plugin.podIPs[id] = ip4.String()
+	// add the ip to tracked ips
+	if ipv4 != nil {
+		plugin.addPodIP(id, ipv4.String())
+	}
+	if ipv6 != nil {
+		plugin.addPodIP(id, ipv6.String())
+	}
+
+	if err := plugin.addTrafficShaping(id, annotations); err != nil {
+		return err
+	}
 
-	// The first SetUpPod call creates the bridge; get a shaper for the sake of initialization
-	// TODO: replace with CNI traffic shaper plugin
+	return plugin.addPortMapping(id, name, namespace)
+}
+
+// The first SetUpPod call creates the bridge; get a shaper for the sake of initialization
+// TODO: replace with CNI traffic shaper plugin
+func (plugin *kubenetNetworkPlugin) addTrafficShaping(id kubecontainer.ContainerID, annotations map[string]string) error {
 	shaper := plugin.shaper()
 	ingress, egress, err := bandwidth.ExtractPodBandwidthResources(annotations)
 	if err != nil {
 		return fmt.Errorf("Error reading pod bandwidth annotations: %v", err)
 	}
+	iplist, exists := plugin.podIPs[id]
+	if !exists {
+		return fmt.Errorf("pod %s does not have recorded ips", id)
+	}
+
 	if egress != nil || ingress != nil {
-		if err := shaper.ReconcileCIDR(fmt.Sprintf("%s/32", ip4.String()), egress, ingress); err != nil {
-			return fmt.Errorf("Failed to add pod to shaper: %v", err)
+		for _, ip := range iplist.UnsortedList() {
+			mask := 32
+			if netutils.IsIPv6String(ip) {
+				mask = 128
+			}
+			if err != nil {
+				return fmt.Errorf("failed to setup traffic shaping for pod ip%s", ip)
+			}
+
+			if err := shaper.ReconcileCIDR(fmt.Sprintf("%v/%v", ip, mask), egress, ingress); err != nil {
+				return fmt.Errorf("Failed to add pod to shaper: %v", err)
+			}
 		}
 	}
+	return nil
+}
 
-	// TODO: replace with CNI port-forwarding plugin
+// TODO: replace with CNI port-forwarding plugin
+func (plugin *kubenetNetworkPlugin) addPortMapping(id kubecontainer.ContainerID, name, namespace string) error {
 	portMappings, err := plugin.host.GetPodPortMappings(id.ID)
 	if err != nil {
 		return err
 	}
-	if portMappings != nil && len(portMappings) > 0 {
-		if err := plugin.hostportManager.Add(id.ID, &hostport.PodPortMapping{
+
+	if nil == portMappings || 0 == len(portMappings) {
+		return nil
+	}
+
+	iplist, exists := plugin.podIPs[id]
```

Comment:
Nothing holds the plugin.mu.Lock() around addPortMapping() anymore (since you removed it for setup()) so now you need to protect against concurrent accesses to plugin.podIPs here.